core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java

```diff
@@ -0,0 +1,257 @@
+        AvroSchemaUtil.fromOption(record.getSchema().getField("point").schema()).getElementType()));
+    pointRecord.put("x", 1);
+    pointRecord.put("y", 2);
+    record.put("point", ImmutableList.of(pointRecord));
+
+    NameMapping nameMapping = MappingUtil.create(new Schema(
+        // Optional array field missing.
+        Types.NestedField.required(0, "id", Types.LongType.get())));
+
+    Schema readSchema = new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(22, "point",
+            Types.ListType.ofOptional(21, Types.StructType.of(
+                Types.NestedField.required(19, "x", Types.IntegerType.get()),
+                Types.NestedField.optional(18, "y", Types.IntegerType.get())
+            ))
+        )
+    );
+
+    Record projected = writeAndRead(writeSchema, readSchema, record, nameMapping);
+    Assert.assertNotNull("Field missing from table mapping is renamed", projected.getSchema().getField("point_r22"));
+    Assert.assertNull("point field is not projected", projected.get("point_r22"));
+    Assert.assertEquals(34L, projected.get("id"));
+
+    // point array is partially projected
+    nameMapping = MappingUtil.create(new Schema(
+        Types.NestedField.required(0, "id", Types.LongType.get()),
+        Types.NestedField.optional(22, "point",
+            Types.ListType.ofOptional(21, Types.StructType.of(
+                Types.NestedField.required(19, "x", Types.IntegerType.get())))
+        )
+    ));
+
+    projected = writeAndRead(writeSchema, readSchema, record, nameMapping);
+    Record point = ((List<Record>) projected.get("point")).get(0);
+
+    Assert.assertNotNull("Field missing from table mapping is renamed", point.getSchema().getField("y_r18"));
+    Assert.assertEquals("point.x is projected", 1, point.get("x"));
+    Assert.assertNull("point.y is not projected", point.get("y_r18"));
+    Assert.assertEquals(34L, projected.get("id"));
+  }
+
+  @Test
+  public void testAliases() throws IOException {
+    Schema writeSchema = new Schema(
+        Types.NestedField.optional(22, "points",
+            Types.ListType.ofOptional(21, Types.StructType.of(
+                Types.NestedField.required(19, "x", Types.IntegerType.get())))));
+
+    Record record = new Record(AvroSchemaUtil.convert(writeSchema, "table"));
+    Record pointRecord = new Record(AvroSchemaUtil.fromOption(
+        AvroSchemaUtil.fromOption(record.getSchema().getField("points").schema()).getElementType()));
+    pointRecord.put("x", 1);
+    record.put("points", ImmutableList.of(pointRecord));
+
+    NameMapping nameMapping = NameMapping.of(
+        MappedFields.of(
+            MappedField.of(22, "points", MappedFields.of(
+                MappedField.of(21, "element", MappedFields.of(
+                    // Name mapping has aliases for field x
+                    MappedField.of(19, Lists.newArrayList("x", "y", "z"))))))));
```

Comment:
I don't think that y or z is required here. The ID and x are the only two pieces of information needed for the rename.